### PR TITLE
refactor: ArgillaTrainer for unified variable usage

### LINKED ISSUE #4029: [FEATURE] refactor `ArgillaTrainer`
**Is your feature request related to a problem? Please describe.**
The `ArgillaTrainer` and reusability can be improved.

**Describe the solution you'd like**
- schemas for args, return types, default args
- clearer distinction model_args, training_args
- methods to get model and training args
- aligning naming across frameworks

**Describe alternatives you've considered**
N.A.

**Additional context**
N.A.

### PR DESCRIPTION
**Pull request by davidberenstein1957**
# Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Closes #4029 

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [ ] New feature (non-breaking change which adds functionality)
- [X] Refactor (change restructuring the codebase without changing functionality)
- [ ] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

N.A.

**Checklist**

- [ ] I added relevant documentation
- [X] I followed the style guidelines of this project
- [X] I did a self-review of my code
- [X] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [X] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

### CODE REVIEW

**Comment by davidberenstein1957** [comment]
Look great @sdiazlor. Let's try to review and align this with the website updates.

File: src/argilla/client/feedback/training/base.py [hunk 1]
```diff
@@ -347,8 +345,27 @@ def push_to_huggingface(self, repo_id: str, generate_card: Optional[bool] = True
...
+        args = tuple(new_args)
+
+        new_kwargs = {}
+        for key, value in kwargs.items():
+            if hasattr(value, "__fields__"):
+                new_kwargs.update(value.dict())
+            else:
+                new_kwargs[key] = value
+        kwargs = new_kwargs
+        return super().update_config(*args, **kwargs)
```
Comment (plaguss) [inline src/argilla/client/feedback/training/base.py]: Just a reminder here. When we migrate to [pydantic 2](https://docs.pydantic.dev/latest/migration/#changes-to-pydanticbasemodel) we should update this method.

File: src/argilla/client/feedback/training/schemas/args.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import TYPE_CHECKING
+
+from pydantic import BaseModel, ConfigDict
+
+if TYPE_CHECKING:
+    import sentence_transformers
```
Comment (plaguss) [inline src/argilla/client/feedback/training/schemas/args.py]: Where are we using this?

File: src/argilla/client/feedback/training/schemas/base.py [hunk 3]
```diff
@@ -39,43 +53,45 @@
...
+        return [Framework(name) for name in self._supported_frameworks_names]
 
+    def __repr__(self) -> str:
+        return textwrap.dedent(
+            f"{self.__class__.__name__}" f"formatting_func:\n{self.formatting_func}, "
+            if self.formatting_func
+            else "" f"defaults:\n{self.defaults}), "
+            if self.defaults
+            else ""
+        )
```
Comment (plaguss) [inline src/argilla/client/feedback/training/schemas/base.py]: How do we expect this to render? I see the result a bit weird:
```python
textwrap.dedent(
    f"{'class_name'}" f"formatting_func:\n{'value'}, "
    if 'value'
    else "" f"defaults:\n{'defaults'}), "
    if 'defaults'
    else ""
)
# >>> 'class_nameformatting_func:\nvalue, '
```

File: src/argilla/client/feedback/training/schemas/base.py [hunk 4]
```diff
@@ -407,13 +464,13 @@ def for_chat_completion(
             [Dict[str, Any]], Union[None, Tuple[str, str, str, str], Iterator[Tuple[str, str, str, str]]]
         ],
     ) -> "TrainingTaskForChatCompletion":
-        """Training data for chat completion
+        """Training data for chat comletion
```
Comment (plaguss) [inline src/argilla/client/feedback/training/schemas/base.py]: typo here

File: src/argilla/client/feedback/training/schemas/base.py [hunk 5]
```diff
@@ -499,7 +557,7 @@ def for_sentence_similarity(
                 List[Dict[str, str]],
             ],
         ] = None,
-        label_strategy: Optional[Union[LabelQuestionUnification, RatingQuestionUnification]] = None,
+        label_strategy: Optional[LabelQuestionUnification] = None,
```
Comment (plaguss) [inline src/argilla/client/feedback/training/schemas/base.py]: We decided to support here both `LabelQuestionUnification` and `RatingQuestionUnification` I think?

File: src/argilla/client/feedback/training/schemas/base.py [hunk 6]
```diff
@@ -573,22 +632,12 @@ def for_sentence_similarity(
                     _LOGGER.info(f"No label strategy defined. Using default strategy for {type(label)}.")
                 if isinstance(label, LabelQuestion):
                     label = LabelQuestionUnification(**unification_kwargs)
-                elif isinstance(label, RatingQuestion):
-                    label = RatingQuestionUnification(**unification_kwargs)
+                elif isinstance(label, RankingQuestion):
+                    label = RankingQuestionUnification(**unification_kwargs)
```
Comment (plaguss) [inline src/argilla/client/feedback/training/schemas/base.py]: It should be `RatingQuestion` as per https://github.com/argilla-io/argilla/pull/4171

File: src/argilla/client/feedback/training/schemas/base.py [hunk 7]
```diff
@@ -1341,23 +1281,15 @@ def _prepare_for_training_with_transformers(
...
-    """
-    Union[Tuple[str, str, str, str], List[Tuple[str, str, str, str]]]
-    """
-
-    format: Union[Tuple[str, str, str, str], List[Tuple[str, str, str, str]]]
-
-
 class TrainingTaskForChatCompletion(BaseModel, TrainingData):
-    """Training data for chat completion
+    """Training data for chat comletion
```
Comment (plaguss) [inline src/argilla/client/feedback/training/schemas/base.py]: typo here

File: src/argilla/client/feedback/training/schemas/base.py [hunk 8]
```diff
@@ -660,6 +699,14 @@ def __label2id__(self):
     def __id2label__(self):
         return self.label.question.__id2label__
 
+    @property
+    def label(self):
+        return self.defaults.label
+
+    @property
+    def text(self):
+        return self.defaults.text
```
Comment (plaguss) [inline src/argilla/client/feedback/training/schemas/base.py]: Maybe we could add some typing hints here

File: src/argilla/client/feedback/training/schemas/base.py [hunk 9]
```diff
@@ -1250,11 +1192,22 @@ class TrainingTaskForQuestionAnswering(BaseModel, TrainingData):
...
+    def question(self):
+        return self.defaults.question
+
+    @property
+    def context(self):
+        return self.defaults.context
+
+    @property
+    def answer(self):
+        return self.defaults.answer
```
Comment (plaguss) [inline src/argilla/client/feedback/training/schemas/base.py]: Maybe we could add some type hints here

File: src/argilla/client/feedback/training/schemas/base.py [hunk 10]
```diff
@@ -1511,14 +1406,13 @@ def __id2label__(self):
...
-            f"\n\t all_labels={self.__all_labels__}"
-            f"\n\t formatting_funct={self.formatting_func}"
-        )
+    @property
+    def label(self):
+        return self.defaults.label
+
+    @property
+    def texts(self):
+        return self.defaults.texts
```
Comment (plaguss) [inline src/argilla/client/feedback/training/schemas/base.py]: Maybe we could add some type hints here

File: src/argilla/client/feedback/training/schemas/base.py [hunk 11]
```diff
@@ -1558,24 +1446,23 @@ def _format_data(self, dataset: "FeedbackDataset") -> List[Dict[str, Any]]:
...
-                    else:
-                        value = example[k]
-
+                        if value.lstrip("-").isdigit():
+                            value = int(value)
+                        else:
+                            value = float(value)
+                        if isinstance(self.label, RankingQuestionUnification):
+                            max_value = max([float(x) for x in self.label.question.__all_labels__])
+                            value = (value / 100) * float(max_value)
```
Comment (plaguss) [inline src/argilla/client/feedback/training/schemas/base.py]: This was updated with https://github.com/argilla-io/argilla/pull/4171 to use the `RatingQuestionUnification`

File: src/argilla/client/feedback/training/base.py (hunk 1 above)
Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/base.py]: @plaguss, I also decided to remove this and replace it with a more simplified warning.

File: src/argilla/client/feedback/training/schemas/args.py (hunk 2 above)
Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/schemas/args.py]: I deleted this too.

File: src/argilla/client/feedback/training/schemas/base.py (hunk 3 above)
Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/schemas/base.py]: Thanks, this should be updated

File: src/argilla/client/feedback/training/schemas/base.py (hunk 5 above)
Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/schemas/base.py]: check

File: src/argilla/client/feedback/training/schemas/base.py (hunk 6 above)
Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/schemas/base.py]: check

File: src/argilla/client/feedback/training/schemas/base.py (hunk 10 above)
Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/schemas/base.py]: check

File: src/argilla/client/feedback/training/schemas/base.py (hunk 8 above)
Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/schemas/base.py]: check

File: src/argilla/client/feedback/training/schemas/base.py (hunk 9 above)
Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/schemas/base.py]: check

File: src/argilla/client/feedback/training/schemas/base.py (hunk 11 above)
Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/schemas/base.py]: check

Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/schemas/base.py]: check

File: src/argilla/client/feedback/training/schemas/base.py (hunk 10 above)
Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/schemas/base.py]: check

**Pull request merged**